sycl/plugins/level_zero/pi_level0.cpp

```diff
@@ -366,31 +365,86 @@ pi_result _pi_device::initialize() {
   return PI_SUCCESS;
 }
 
-// Crate a new command list to be used in a PI call
+// Retrieve an available command list to be used in a PI call
 pi_result
-_pi_device::createCommandList(ze_command_list_handle_t *ZeCommandList) {
-  // Create the command list, because in L0 commands are added to
+_pi_device::getAvailableCommandList(ze_command_list_handle_t *ZeCommandList,
+                                    ze_command_queue_handle_t ZeCommandQueue,
+                                    ze_fence_handle_t *ZeFence) {
+  // Create/Reuse the command list, because in L0 commands are added to
   // the command lists, and later are then added to the command queue.
-  //
-  // TODO: Figure out how to lower the overhead of creating a new list
-  // for each PI command, if that appears to be important.
+  // Each command list is paired with an associated fence to track when the
+  // command list is available for reuse.
+  _pi_result pi_result = PI_OUT_OF_RESOURCES;
   ze_command_list_desc_t ZeCommandListDesc = {};
   ZeCommandListDesc.version = ZE_COMMAND_LIST_DESC_VERSION_CURRENT;
+  ze_fence_desc_t ZeFenceDesc = {};
+
+  // Initally, we need to check if a command list has already been created
+  // on this device that is available for use. If so, then reuse that
+  // L0 Command List and Fence for this PI call.
+  if (this->ZeCommandListCache.size() > 0) {
+    this->ZeCommandListCacheMutex.lock();
+    *ZeCommandList = this->ZeCommandListCache.front();
+    *ZeFence = this->ZeCommandListFenceMap[*ZeCommandList];
+    this->ZeCommandListCache.pop_front();
+    this->ZeCommandListCacheMutex.unlock();
+    return PI_SUCCESS;
+  }
 
-  // TODO: can we just reset the command-list created when an earlier
-  // command was submitted to the queue?
-  ZE_CALL(zeCommandListCreate(ZeDevice, &ZeCommandListDesc, ZeCommandList));
+  // If there are no available command lists in the cache, then we check for
+  // command lists that have already signalled, but have not been added to the
+  // available list yet. Each command list has a fence associated which tracks
+  // if a command list has completed dispatch of its commands and is ready for
+  // reuse. If a command list is found to have been signalled, then the
+  // command list & fence are reset and we return.
+  ze_result_t result = ZE_RESULT_ERROR_UNKNOWN;
+  this->ZeCommandListFenceMapMutex.lock();
+  std::map<ze_command_list_handle_t, ze_fence_handle_t>::iterator it =
+      this->ZeCommandListFenceMap.begin();
+  while (it != this->ZeCommandListFenceMap.end()) {
+    result = zeFenceQueryStatus(it->second);
+    if (result == ZE_RESULT_SUCCESS) {
+      ZE_CALL(zeCommandListReset(it->first));
```

Comment:
Use `ZE_CALL_NOCHECK` and re-use the similar code in 2 other places. Maybe create a method in pi_device that takes an optional `event` to consider (or iterates the entire ZeCommandListFenceMap).